graylog2-server/src/main/java/org/graylog/plugins/views/search/Query.java

```diff
@@ -231,6 +251,57 @@ public Query addStreamsToFilter(Set<String> streamIds) {
         return toBuilder().filter(newFilter).build();
     }
 
+    public Query replaceStreamCategoryFilters(Function<Collection<String>, Stream<String>> categoryMappingFunction,
+                                              StreamPermissions streamPermissions) {
+        if (filter() == null) {
+            return this;
+        }
+        return toBuilder()
+                .filter(streamCategoryToStreamFiltersRecursively(filter(), categoryMappingFunction, streamPermissions))
+                .build();
+    }
+
+    private Filter streamCategoryToStreamFiltersRecursively(Filter filter,
+                                                            Function<Collection<String>, Stream<String>> categoryMappingFunction,
+                                                            StreamPermissions streamPermissions) {
+        if (filter.filters() == null || filter.filters().isEmpty()) {
+            return filter;
+        }
+        Set<Filter> mappedFilters = new HashSet<>();
+        for (Filter f : filter.filters()) {
+            Filter mappedFilter = streamCategoryToStreamFilter(f, categoryMappingFunction, streamPermissions);
+            if (mappedFilter != null) {
+                mappedFilter = streamCategoryToStreamFiltersRecursively(mappedFilter, categoryMappingFunction, streamPermissions);
+                mappedFilters.add(mappedFilter);
+            }
+        }
+        if (mappedFilters.isEmpty()) {
+            return null;
+        }
+        return filter.toGenericBuilder().filters(mappedFilters.stream().filter(Objects::nonNull).collect(toSet())).build();
+    }
+
+    private Filter streamCategoryToStreamFilter(Filter filter,
+                                                Function<Collection<String>, Stream<String>> categoryMappingFunction,
+                                                StreamPermissions streamPermissions) {
+        if (filter instanceof StreamCategoryFilter scf) {
+            String[] mappedStreamIds = categoryMappingFunction.apply(List.of(scf.category()))
+                    .filter(streamPermissions::canReadStream)
+                    .toArray(String[]::new);
+            if (mappedStreamIds.length == 0) {
+                return null;
+            }
+            // Replace the category with an OrFilter of stream IDs and then add any filters form the original filter
+            Filter streamFilter = StreamFilter.anyIdOf(mappedStreamIds).toGenericBuilder().build();
+            if (filter.filters() != null) {
+                streamFilter = streamFilter.toGenericBuilder().filters(filter.filters()).build();
+            }
+            return streamFilter;
```

Comment:
Could this piece of code be moved to `StreamCategoryFilter` class?
It seems like it is a little bit more of a `StreamCategoryFilter` responsibility to implement its mapping to a stream filter and Query class could be released from a need to implement this functionality.

Reply:
There is a potential it could slightly simplify the tests - you could have an unit test for StreamCategoryFilter testing its mapping and try to mock it in QueryTest.

Reply:
I moved the code over to `StreamCategoryFilter`, created unit tests, and simplified the `QueryTest` logic in my latest commit. Thanks for the feedback!